### PR TITLE
Imag for tf.raw_ops

### PR DESCRIPTION
**Pull request by Enigmafiend**
Close https://github.com/unifyai/ivy/issues/17595

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by KareemMAX** [comment]
To streamline the review process, can you make a comment and write "add_frontend_checklist" (without double quotes). Our bots will edit your comment to a checklist you need to follow, please tick the points you have completed so I can review it

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by Enigmafiend** [comment]
add_frontend_checklist

**Comment by Enigmafiend** [comment]
removed a print statement and a supported dtype mislabel

**Comment by KareemMAX** [comment]
add_frontend_checklist

**Comment by KareemMAX** [comment]
@Enigmafiend seems like our automation didn't pickup the comment because you added a new line 😅 would you mind making another comment without that newline? Just "add_frontend_checklist". After that please fill the checklist accordingly

**Comment by Enigmafiend** [comment]
add_frontend_checklist

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py [hunk 1]
```diff
@@ -3917,3 +3917,32 @@ def test_tensorflow_Zeta(
...
+):
+    input_dtype, x = dtype_and_x
+    helpers.test_frontend_function(
+        input_dtypes=input_dtype,
+        test_flags=test_flags,
+        frontend=frontend,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        rtol=1e-2,
+        atol=1e-2,
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3945]: Are those necessary?

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Seems like you missed `Tout` argument. I'm not really sure if that should be considered as `raw_ops_alias` or not. That's why I didn't go through the majority of the checklist. I'll ask internally and get back to you.

Also, please consider using `pre-commit` to keep your code linted, I'll use our lint-bot to lint your code for now

**Comment by KareemMAX** [comment]
ivy-gardener

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py (hunk 1 above)
Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3946]: Not really, the test runs perfectly fine without it but I was going through the test_math.py and the test for the standard imag had one included and from an engineering pov it kind of makes sense to declare some tolerance for extreme edge cases so I transposed the set from the test_math.py

**Comment by KareemMAX** [comment]
@Enigmafiend following up on the `map_raw_ops_alias`, sadly your function doesn't qualify as an alias in that case. So you must rewrite it as a composite function so we can merge it.

Once you finish with this change please let me know by re-requesting review on github

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py (hunk 1 above)
Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3946]: As for the first issue, I conferred with a Ivy Engineer and they suggested that I use the map_raw_ops_alias and given that the math.imag does not require or pass a Tout argument I chose to neglect it, if it is to be included do let me know

Comment (KareemMAX) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3946]: We know do use safety factor instead, also I'm thinking this tolerance is really high. We will remove it from our other tests as well

**Comment by Enigmafiend** [comment]
ivy-gardener

**Comment by Enigmafiend** [comment]
add_frontend_checklist

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py [hunk 2]
```diff
@@ -3917,3 +3917,38 @@ def test_tensorflow_Zeta(
...
+    frontend,
+    test_flags,
+    fn_tree,
+    on_device,
+):
+    input_dtype, xs = dtype_and_xs
+    if input_dtype[0] == "complex128":
+        Tout = "float64"
+    elif input_dtype[0] == "complex64":
+        Tout = "float32" if Tout else None
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3944]: This is against convention, also you should generate `Tout` from another strategy as per tf documentation, `Tout` is decoupled from `input`

File: ivy/functional/frontends/tensorflow/raw_ops.py [hunk 3]
```diff
@@ -870,3 +870,24 @@ def Prod(*, input, axis, keep_dims=False, name="Prod"):
...
+
+
+Imag = to_ivy_arrays_and_back(
+    with_supported_dtypes(
+        {
+            "2.12.0 and below": "complex64",
+        },
+        "tensorflow",
+    )(map_raw_ops_alias(tf_frontend.math.imag))
+)
```
Comment (KareemMAX) [inline ivy/functional/frontends/tensorflow/raw_ops.py:882]: This is no longer necessary, and actually a redeclaration that is problematic and we can't merge with it present

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py [hunk 4]
```diff
@@ -3917,3 +3917,38 @@ def test_tensorflow_Zeta(
...
     )
+
+
+# Imag
+@handle_frontend_test(
+    fn_tree="tensorflow.raw_ops.Imag",
+    dtype_and_xs=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("complex"),
+    ),
+    Tout=st.booleans(),
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3928]: I guess that is wrong. You should sample from `float32` and `float64` instead

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @Enigmafiend

I've addressed some issues with your PR that should be fixed before merging your PR. Please take a look I've also updated the checklist accordingly. 

After you finish please request a review from me again. Thanks 😄🚀

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py (hunk 4 above)
Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3928]: This shouldnt be wrong as tf as far as ik takes only complex inputs 
image

Comment (KareemMAX) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3928]: I meant for `Tout`, as per our conversation privately. You should remove reference of `Tout` completely from here.

**Comment by Enigmafiend** [comment]
ivy-gardener

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py (hunk 4 above)
Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py]: I have used the test code for tf.raw_ops.Angle as it is because 1.It works and checks all the boxes included in the deep dive docs and 2. They are more or less the same function (Both take a complex input and one returns the Imaginary component while the other returns the angle from the real axis)

**Comment by Enigmafiend** [comment]
ivy-gardener

**Comment by Enigmafiend** [comment]
Another point to note that I missed to mention when we talked, is that the Tout coupling makes sense as a complex 64 is practically 2 Float 32 Elements combined into one (Hence the increased memory as 32 bits for the Real element and 32 for the Imaginary) therefore trying to assign it to a Tout of float64 is an invalid means of memory allocation and not accepted by the cpu setup

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py [hunk 5]
```diff
@@ -3917,3 +3917,38 @@ def test_tensorflow_Zeta(
...
+    frontend,
+    test_flags,
+    fn_tree,
+    on_device,
+):
+    input_dtype, xs = dtype_and_xs
+    if input_dtype[0] == "complex128":
+        Tout = "float64"
+    elif input_dtype[0] == "complex64":
+        Tout = "float32" if Tout else None
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3944]: A slight renaming would be better. Because it's counter-intuitive to use `Tout` for 2 different uses. Maybe the test parameter one should be named `send_type` or something similar.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py [hunk 6]
```diff
@@ -3917,3 +3917,38 @@ def test_tensorflow_Zeta(
         x=x[0],
         q=x[1],
     )
+
+
+# Imag
+@handle_frontend_test(
+    fn_tree="tensorflow.raw_ops.Imag",
+    dtype_and_xs=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("complex"),
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3926]: Can you change this to `"valid"` and use `@with_supported_dtypes`/`@with_unsupported_dtype`

**Review by KareemMAX — CHANGES_REQUESTED** [review]
I've pointed out some issues. Your PR is really near completion. Thank you so much @Enigmafiend 😄

**Comment by Enigmafiend** [comment]
ivy-gardener

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py [hunk 7]
```diff
@@ -3917,3 +3917,38 @@ def test_tensorflow_Zeta(
...
+    ),
+    Tout=st.sampled_from(["float64", "float32"]),
```
[the remark this answers is not in the document]
Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py]: This works the same as st.boolean() since Tout will be coupled anyways but makes more intuitive sense when trying to understand the test

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py [hunk 8]
```diff
@@ -3917,3 +3917,38 @@ def test_tensorflow_Zeta(
...
     )
+
+
+# Imag
+@handle_frontend_test(
+    fn_tree="tensorflow.raw_ops.Imag",
+    dtype_and_xs=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("valid"),
+    ),
+    send_Tout=st.sampled_from(["float64", "float32"]),
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_raw_ops.py:3928]: I think that still should be `st.booleans()`. Right?

**Review by KareemMAX — APPROVED** [review]
LGTM! Merging now
